File: app/lib/models/log_entry.dart
/// Dart equivalents of the v2 StoredEntry schema.
///
/// Field names use camelCase per Dart convention; JSON keys use snake_case
/// matching the wire protocol.
library;

import 'data_state_models.dart';
import 'exception_models.dart';
import 'log_enums.dart';
import 'log_sub_models.dart';

export 'data_state_models.dart';
export 'exception_models.dart';
export 'log_enums.dart';
export 'log_sub_models.dart';

// ─── LogEntry ────────────────────────────────────────────────────────

class LogEntry {
  final String id;
  final String timestamp;
  final String sessionId;
  final EntryKind kind;
  final Severity severity;

  final String? message;
  final String? tag;
  final ExceptionData? exception;
  final String? parentId;
  final String? groupId;
  final String? prevId;
  final String? nextId;
  final WidgetPayload? widget;
  final bool replace;
  final IconRef? icon;
  final Map<String, String>? labels;
  final String? generatedAt;
  final String? sentAt;

  final String? key;
  final dynamic value;
  final bool override_;
  final DisplayLocation display;

  final SessionAction? sessionAction;
  final ApplicationInfo? application;
  final Map<String, dynamic>? metadata;

  final String? receivedAt;

  const LogEntry({
    required this.id,
    required this.timestamp,
    required this.sessionId,
    required this.kind,
    required this.severity,
    this.message,
    this.tag,
    this.exception,
    this.parentId,
    this.groupId,
    this.prevId,
    this.nextId,
    this.widget,
    this.replace = false,
    this.icon,
    this.labels,
    this.generatedAt,
    this.sentAt,
    this.key,
    this.value,
    this.override_ = true,
    this.display = DisplayLocation.defaultLoc,
    this.sessionAction,
    this.application,
    this.metadata,
    this.receivedAt,
  });

  factory LogEntry.fromJson(Map<String, dynamic> json) {
    return LogEntry(
      id: json['id'] as String,
      timestamp: json['timestamp'] as String,
      sessionId: json['session_id'] as String,
      kind: parseEntryKind(json['kind'] as String),
      severity: parseSeverity(json['severity'] as String? ?? 'info'),
      message: json['message'] as String?,
      tag: json['tag'] as String?,
      exception: json['exception'] != null
          ? ExceptionData.fromJson(json['exception'] as Map<String, dynamic>)
          : null,
      parentId: json['parent_id'] as String?,
      groupId: json['group_id'] as String?,
      prevId: json['prev_id'] as String?,
      nextId: json['next_id'] as String?,
      widget: json['widget'] != null
          ? WidgetPayload.fromJson(json['widget'] as Map<String, dynamic>)
          : null,
      replace: json['replace'] as bool? ?? false,
      icon: json['icon'] != null
          ? IconRef.fromJson(json['icon'] as Map<String, dynamic>)
          : null,
      labels: (json['labels'] as Map<String, dynamic>?)?.map(
        (k, v) => MapEntry(k, v as String),
      ),
      generatedAt: json['generated_at'] as String?,
      sentAt: json['sent_at'] as String?,
      key: json['key'] as String?,
      value: json['value'],
      override_: json['override'] as bool? ?? true,
      display: parseDisplayLocation(json['display'] as String? ?? 'default'),
      sessionAction: parseSessionAction(json['session_action'] as String?),
      application: json['application'] != null
          ? ApplicationInfo.fromJson(
              json['application'] as Map<String, dynamic>,
            )
          : null,
      metadata: json['metadata'] as Map<String, dynamic>?,
      receivedAt: json['received_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'timestamp': timestamp,
    'session_id': sessionId,
    'kind': kind.name,
    'severity': severity.name,
    if (message != null) 'message': message,
    if (tag != null) 'tag': tag,
    if (exception != null) 'exception': exception!.toJson(),
    if (parentId != null) 'parent_id': parentId,
    if (groupId != null) 'group_id': groupId,
    if (prevId != null) 'prev_id': prevId,
    if (nextId != null) 'next_id': nextId,
    if (widget != null) 'widget': widget!.toJson(),
    'replace': replace,
    if (icon != null) 'icon': icon!.toJson(),
    if (labels != null) 'labels': labels,
    if (generatedAt != null) 'generated_at': generatedAt,
    if (sentAt != null) 'sent_at': sentAt,
    if (key != null) 'key': key,
    if (value != null) 'value': value,
    'override': override_,
    'display': switch (display) {
      DisplayLocation.defaultLoc => 'default',
      DisplayLocation.static_ => 'static',
      DisplayLocation.shelf => 'shelf',
    },
    if (sessionAction != null) 'session_action': sessionAction!.name,
    if (application != null) 'application': application!.toJson(),
    if (metadata != null) 'metadata': metadata,
    if (receivedAt != null) 'received_at': receivedAt,
  };
}

File: shared/src/log-entry.ts
import { z } from 'zod';

// ─── Enums ───────────────────────────────────────────────────────────

export const Severity = z.enum([
  'debug',
  'info',
  'warning',
  'error',
  'critical',
]);
export type Severity = z.infer<typeof Severity>;

export const LogType = z.enum([
  'text',
  'json',
  'html',
  'binary',
  'image',
  'state',
  'group',
  'rpc',
  'session',
  'custom',
]);
export type LogType = z.infer<typeof LogType>;

export const GroupAction = z.enum(['open', 'close']);
export type GroupAction = z.infer<typeof GroupAction>;

export const SessionAction = z.enum(['start', 'end', 'heartbeat']);
export type SessionAction = z.infer<typeof SessionAction>;

export const RpcDirection = z.enum(['request', 'response', 'error']);
export type RpcDirection = z.infer<typeof RpcDirection>;

// ─── Sub-schemas ─────────────────────────────────────────────────────

export const SourceLocation = z.object({
  /** File path or URI */
  uri: z.string(),
  /** 1-based line number */
  line: z.number().int().optional(),
  /** 1-based column number */
  column: z.number().int().optional(),
  /** Function/method name */
  symbol: z.string().optional(),
});
export type SourceLocation = z.infer<typeof SourceLocation>;

export const StackFrame = z.object({
  location: SourceLocation,
  /** true for node_modules, dart:, etc. */
  isVendor: z.boolean().optional(),
  /** Original unparsed frame string */
  raw: z.string().optional(),
});
export type StackFrame = z.infer<typeof StackFrame>;

export const ExceptionData: z.ZodType<{
  type?: string;
  message: string;
  stackTrace?: Array<{ location: { uri: string; line?: number; column?: number; symbol?: string }; isVendor?: boolean; raw?: string }>;
  cause?: unknown;
}> = z.object({
  /** e.g., "TypeError", "FormatException" */
  type: z.string().optional(),
  message: z.string(),
  stackTrace: z.array(StackFrame).optional(),
  /** Chained exception cause */
  cause: z.lazy(() => ExceptionData).optional(),
});
export type ExceptionData = z.infer<typeof ExceptionData>;

export const ApplicationInfo = z.object({
  /** e.g., "my-api-server" */
  name: z.string(),
  /** e.g., "1.2.3" */
  version: z.string().optional(),
  /** e.g., "development" */
  environment: z.string().optional(),
});
export type ApplicationInfo = z.infer<typeof ApplicationInfo>;

export const IconRef = z.object({
  /** Iconify ID, e.g., "mdi:home" */
  icon: z.string(),
  /** Hex color override, e.g., "#FF0000" */
  color: z.string().optional(),
  /** Size in dp, default 14 */
  size: z.number().optional(),
});
export type IconRef = z.infer<typeof IconRef>;

export const ImageData = z
  .object({
    /** Base64-encoded image data (for inline transport) */
    data: z.string().optional(),
    /** Reference ID returned by POST /api/v1/upload */
    ref: z.string().optional(),
    /** MIME type, e.g., "image/png" */
    mimeType: z.string().optional(),
    /** Display label */
    label: z.string().optional(),
    /** Width in pixels (metadata) */
    width: z.number().int().optional(),
    /** Height in pixels (metadata) */
    height: z.number().int().optional(),
  })
  .refine((d) => d.data !== undefined || d.ref !== undefined, {
    message: 'ImageData requires either data (base64) or ref (upload reference)',
  });
export type ImageData = z.infer<typeof ImageData>;

// ─── LogEntry (the unified protocol message) ────────────────────────

export const LogEntry = z.object({
  // ── Required fields ──
  /** Client-assigned ID. String to allow both UUID and custom string IDs. */
  id: z.string(),
  /** ISO 8601 timestamp with millisecond precision */
  timestamp: z.string().datetime({ offset: true }),
  /** Session ID — groups logs from a single application run */
  session_id: z.string(),
  /** Log severity level */
  severity: Severity,
  /** Discriminator for log entry type */
  type: LogType,

  // ── Application metadata (optional) ──
  application: ApplicationInfo.optional(),

  // ── Section targeting ──
  /** Which UI section this log belongs to. Defaults to "events" */
  section: z.string().optional(),

  // ── Content fields (type-dependent) ──
  /** For type: "text" — plain text content */
  text: z.string().optional(),
  /** For type: "json" — arbitrary JSON payload */
  json: z.unknown().optional(),
  /** For type: "html" — HTML string content */
  html: z.string().optional(),
  /** For type: "binary" — base64-encoded binary data */
  binary: z.string().optional(),
  /** For type: "image" — image data or reference */
  image: ImageData.optional(),

  // ── Exception / stack trace (can accompany any type) ──
  exception: ExceptionData.optional(),

  // ── Icon reference ──
  icon: IconRef.optional(),

  // ── Group operations ──
  /** Group ID. All entries in a group share this. */
  group_id: z.string().optional(),
  /** What action to perform on the group */
  group_action: GroupAction.optional(),
  /** Group display label (used with group_action: "open") */
  group_label: z.string().optional(),
  /** Whether the group starts collapsed in the viewer */
  group_collapsed: z.boolean().optional(),

  // ── Sticky pinning ──
  /** When true, this entry (or group) pins to the top of the viewport when scrolled past */
  sticky: z.boolean().optional(),

  // ── State operations (type: "state") ──
  /** State key for upsert. Unique per session. */
  state_key: z.string().optional(),
  /** State value. null = delete the key. */
  state_value: z.unknown().optional(),

  // ── Session control (type: "session") ──
  session_action: SessionAction.optional(),

  // ── Ordering hints ──
  /** Insert this log visually after the entry with this ID */
  after_id: z.string().optional(),
  /** Insert this log visually before the entry with this ID */
  before_id: z.string().optional(),

  // ── 2-Way RPC (type: "rpc") ──
  /** Unique RPC call ID */
  rpc_id: z.string().uuid().optional(),
  /** Direction of the RPC message */
  rpc_direction: RpcDirection.optional(),
  /** RPC method name */
  rpc_method: z.string().optional(),
  /** RPC arguments (request direction) */
  rpc_args: z.unknown().optional(),
  /** RPC response data (response direction) */
  rpc_response: z.unknown().optional(),
  /** RPC error message (error direction) */
  rpc_error: z.string().optional(),

  // ── Request timing metadata ──
  /** When the log was originally generated */
  generated_at: z.string().datetime({ offset: true }).optional(),
  /** When the log was sent over the wire */
  sent_at: z.string().datetime({ offset: true }).optional(),

  // ── Tags ──
  /** Arbitrary key-value tags for filtering/searching */
  tags: z.record(z.string(), z.string()).optional(),

  // ── Patch 3: In-place entry updates ──
  /** When true + id matches existing entry, upsert (replace in-place) */
  replace: z.boolean().optional(),

  // ── Custom type fields ──
  /** Custom type discriminator (when type: "custom") */
  custom_type: z.string().optional(),
  /** Custom renderer data (when type: "custom") */
  custom_data: z.unknown().optional(),
});
export type LogEntry = z.infer<typeof LogEntry>;

// ─── Batch Wrapper ───────────────────────────────────────────────────

export const LogBatch = z.object({
  /** Array of log entries */
  entries: z.array(LogEntry).min(1).max(1000),
});
export type LogBatch = z.infer<typeof LogBatch>;
